[RHCLOUD-17832] Internal endpoint for source listing

MikelAlejoBR opened the pull request:
Creates an internal endpoint to be able to list the sources along with their tenants. The code is very similar to any other "List" code, except for:

* The internal function doesn't need a DAO with a `tenantId`, since we're not filtering by that for this endpoint.
* The query just selects the columns that are needed for the response.
* The response only includes the `id`, `availability_status` and `tenant` fields; the ones that `sources-monitor-go` uses.

## Links
[RHCLOUD-17832](https://issues.redhat.com/browse/RHCLOUD-17832)

MikelAlejoBR committed:
feature: internal endpoint for source listing

MikelAlejoBR committed:
feature: OpenApi Spec for the internal endpoints

lpichler commented on dao/source_dao.go:
```diff
@@ -61,6 +61,24 @@ func (s *SourceDaoImpl) List(limit, offset int, filters []util.Filter) ([]m.Sour
 }
 
+func (s *SourceDaoImpl) ListInternal(limit, offset int, filters []util.Filter) ([]m.Source, int64, error) {
+	query := DB.Table("sources").Select("id", "availability_status", "tenant_id")
```
~1. Why only 3 fields ?~
I guess it is because of source-monitor as I saw bellow.

2. We prefer (at least as we are using it everywhere) `DB.Model( &m.Source{})` instead `DB.Table..`

lindgrenj6 replied:
Yeah - since sources-monitor only requires those three fields and this was previously undocumented we're going to lock it down to only the 3 necessary fields, we can always expose more if needbe!

MikelAlejoBR replied:
I changed it to `DB.Model(&m.Source{})` but kept the `.Select` part, so we don't fetch more columns than necessary! Thanks for pointing it out.

lpichler commented on private/openapi-3-v2.0.yaml:
```diff
@@ -0,0 +1,212 @@
+openapi: 3.0.0
```
why we use yaml here as we are using jsons so far(in ruby project, in public) ?

MikelAlejoBR replied:
As I was struggling with editing the spec in JSON, I thought "well, we might as well switch to YAML since it's less painful". I didn't remember that we need to have the spec in JSON format because we're serving it on an endpoint, and for the sake of consistency, I'm transforming the private spec back too to JSON.

lpichler commented on source_handlers_test.go:
```diff
@@ -275,6 +276,85 @@ func TestSourceListSatellite(t *testing.T) {
+		t.Error("not enough objects passed back from DB")
+	}
+
+	for i, src := range out.Data {
```
maybe do we need to sort those `out.Data`  for sure ? function `SortByStringValueOnKey("id", out.Data)` could be used for it. - I don't whether that there is defined order for `out.Data`

MikelAlejoBR replied:
I'm sorry: I've tried to find the function you told me about, but I am not able to find it :eyes: 

Anyway: in this case Gorm seems to be respecting the order we set in the fixtures, so we shouldn't have any problem with the tests. Although we could create a `TODO` for this to be safer. What do you think?

lpichler commented on model/source.go:
```diff
@@ -86,3 +86,17 @@ func (src *Source) ToResponse() *SourceResponse {
+	source := &SourceInternalResponse{
+		Id:                 &id,
+		AvailabilityStatus: &src.AvailabilityStatus.AvailabilityStatus,
+		TenantId:           &tenantId,
```
from request on ruby app - it looks that we are exposing `Tenant.ExternalTenant` for attribute "tenant" in response .

MikelAlejoBR replied:
You are right. I tried it by setting up the Rails app on the local environment and we're returning the external tenant. I changed that.

lindgrenj6 commented on source_handlers.go:
```diff
@@ -325,3 +325,37 @@ func SourceCheckAvailability(c echo.Context) error {
+	if c.Get("cert-auth") != nil {
+		satelliteId := strconv.Itoa(int(dao.Static.GetSourceTypeId("satellite")))
+		filters = append(filters, util.Filter{Name: "source_type_id", Value: []string{satelliteId}})
+	}
```
This check probably isn't necessary for the internal endpoint, we can remove it!

MikelAlejoBR replied:
All right, thanks!

MikelAlejoBR committed:
refactor: use "DB.Model" instead of "DB.Table"

I assumed using "DB.Model" would fetch all the columns and I didn't try
using ".Select" with it. Apparently it is possible, so I'm changing it
to that to be consistent with the rest of the code, and because I rather
have Gorm deduce the table name from the struct than hardcoding the
name —since it might be easy to forget if we ever change the table name—.

MikelAlejoBR committed:
refactor: fix typo

MikelAlejoBR committed:
refactor: output more info when tests don't pass

MikelAlejoBR committed:
refactor: convert OpenApi file to JSON for consistency

The public API's spec is served in an endpoint, as we allow other
services to fetch it if they so require. The internal one doesn't get
exposed, but for consistency, it is a good idea to keep it as JSON as
well.

MikelAlejoBR committed:
refactor: remove unnecessary code

MikelAlejoBR committed:
refactor: return the external tenant instead of the tenant id

The Rails app returned the "external_tenant" of the related tenant of a
given source, so this Go version should basically do the same. In order
to avoid having to fetch the entire table, a "Join" is performed, but
GORM seems not to be able to map stuff directly to an inner struct
—which is a fair limitation, to be honest—.

That's why I'm using a half way solution for this: instead of looping
through the rows and manually mapping the fields, I let GORM handle that
at the expense of fetching all the columns of the joined table, and
letting it make a "LEFT JOIN" instead of an "INNER JOIN".

lindgrenj6 commented on source_handlers.go:
```diff
@@ -325,3 +325,30 @@ func SourceCheckAvailability(c echo.Context) error {
 	return c.JSON(http.StatusAccepted, map[string]interface{}{})
 }
+
+func InternalSourceList(c echo.Context) error {
```
Might I suggest moving this handlerFunc to a different file? Maybe could have an `internal_handlers.go` file - that way we can separate out the 2 internal endpoints there.  

What do you think?

MikelAlejoBR replied:
I like the idea. Applying it now.

lindgrenj6 reviewed (approved):
Looks good, just one tiny not-important thought I had while reviewing it.

MikelAlejoBR committed:
refactor: move internal handler to a separate file

MikelAlejoBR committed:
refactor: fix indentation

MikelAlejoBR committed:
refactor: add missing schema element on the API spec

MikelAlejoBR committed:
refactor: make error example more generic

Instead of giving a "401" as an error example, which may be confusing, a
more generic error helps focusing on what's important: the error
structure.

lpichler reviewed (approved):
LGTM,
@MikelAlejoBR thanks!
